pytest_plugins/fixture_markers.py

```diff
@@ -6,16 +7,40 @@
 
 def pytest_generate_tests(metafunc):
     if 'rhel_contenthost' in metafunc.fixturenames:
-        rhel_parameters = [
-            dict(workflow=settings.content_host.deploy_workflow, rhel_version=ver)
-            for ver in settings.content_host.rhel_versions
-        ]
-        metafunc.parametrize(
-            'rhel_contenthost',
-            rhel_parameters,
-            ids=[f'rhel{r["rhel_version"]}' for r in rhel_parameters],
-            indirect=True,
-        )
+        rhel_parameters = []
+        for ver in settings.content_host.rhel_versions.keys():
```

Comment:
Could iterate over items here to save the dict lookup later

Reply:
Yeah, I can certainly do that. I'm just wondering how to do this so it won't make the code unreadable.
See, I first have to extract various kinds of selectors outside of the loop (new selector kinds can be added by time).
Then, I get to iterate over possible multiple instances of the selectors of the same kind inside the loop and then do the same for other kinds...

Maybe the benefits of keeping the selector-kind-specific in the distinct code blocks might outweight the efficiency benefit of the suggested change?

Reply:
maybe something like this? completely untested and written minutes before i go to bed, so definitely don't take this as a "this will work" suggestion. but what i'm thinking about right now
```python
        matchers = [i.args for i in metafunc.function.pytestmark if i.name == 'rhel_version_list']
        for m in matchers:
            if m:
                list_params = [
                    param
                    for param in settings.content_host.rhel_versions 
                    if str(param['rhel_version']) in str(m[0])
                ]
        # process eventual rhel_nick_match markers
        matchers = [i.args for i in metafunc.function.pytestmark if i.name == 'rhel_version_match']
        for m in matchers:
            if m:
                match_params = [
                    param
                    for param in settings.content_host.rhel_versions
                    if re.fullmatch(m[0], str(param['rhel_version']))
                ]
        rhel_params = []
        for ver in settings.content_host.rhel_versions.keys():
            if ver in set(list_params + match_params):
                # prefer version-specific deploy workflow before using the default one
                workflow = settings.content_host.rhel_versions[ver].get(
                    'deploy_workflow', settings.content_host.deploy_workflow
                )
            rhel_params.append(dict(workflow=workflow, rhel_version=ver))
        if rhel_params:
            metafunc.parametrize(
                'rhel_contenthost',
                rhel_params,
                ids=[f'rhel{r["rhel_version"]}' for r in rhel_params],
                indirect=True,
            )
```

Reply:
the `if m:` checks probably aren't necessary and could save on a bit more visual noise.

Reply:
This looks nice.
However, it changes the behavior of how multiple selector markers are treated.
- if multiple instances of `rhel_version_list` mark are used, only the last processed one will take effect.
- if both kinds of marks are used (`rhel_version_list` together with `rhel_version_match`), they will be combined using `OR` logic, instead of the original `AND`

Am I right?
I don't think any of this is wrong tho, we can go with it.

Reply:
I may be wrong, but I read the original flow as:
```
Get all the versions and data from settings
Reduce that list down to what was specified in rhel_version_list
Reduce that list down to what was specified in rhel_version_match
```
Whereas the altered version is
```
Construct a list of desired versions from rhel_version_list
Construct a list of desired versions from rhel_version_match
Get all the versions and data from settings if they are in either of the above lists
```

Reply:
Now that I look at it more, wouldn't this work better than my previous example?
```python
        matchers = [i.args for i in metafunc.function.pytestmark if i.name == 'rhel_version_list']
        list_params = [
            param
            for param in settings.content_host.rhel_versions 
            if str(param['rhel_version']) in str(matchers)
        ] if matchers else []
        # process eventual rhel_version_match markers
        matchers = [i.args for i in metafunc.function.pytestmark if i.name == 'rhel_version_match']
        match_params = [
            param
            for param in settings.content_host.rhel_versions
            if re.fullmatch(str(matchers), str(param['rhel_version']))
        ] if matchers else []
        rhel_params = []
        for ver, data in settings.content_host.rhel_versions.items():
            if ver in set(list_params + match_params):
                # prefer version-specific deploy workflow before using the default one
                workflow = data.get(
                    'deploy_workflow', settings.content_host.deploy_workflow
                )
            rhel_params.append(dict(workflow=workflow, rhel_version=ver))
        if rhel_params:
            metafunc.parametrize(
                'rhel_contenthost',
                rhel_params,
                ids=[f'rhel{r["rhel_version"]}' for r in rhel_params],
                indirect=True,
            )
```

Reply:
well, I guess, but this does not cover the scenario when no selector marks are passed and thus all versions are supposed to be picked.
I'm increasingly more leaning towards my original solution, tbh.